relayer/provider/matcher.go

```diff
@@ -118,3 +154,47 @@ func isMatchingTendermintClient(a, b tmclient.ClientState) bool {
 func isMatchingTendermintConsensusState(a, b *tmclient.ConsensusState) bool {
 	return reflect.DeepEqual(*a, *b)
 }
+
+// checkTendermintMisbehaviour checks that a proposed consensus state, used to update a tendermint light client,
+// matches the trusted consensus state from the counterparty chain. If there is no cached trusted header then
+// it will be queried from the counterparty. If the consensus states for the proposed header and the trusted header
+// match then both returned values will be nil.
+func checkTendermintMisbehaviour(
+	ctx context.Context,
+	clientID string,
+	proposedHeader *tmclient.Header,
+	cachedHeader IBCHeader,
+	counterparty ChainProvider,
+) (ibcexported.ClientMessage, error) {
+	var (
+		trustedHeader *tmclient.Header
+		err           error
+		ok            bool
+	)
+
+	if cachedHeader == nil {
+		header, err := counterparty.QueryHeaderAtHeight(ctx, proposedHeader.Header.Height)
+		if err != nil {
+			return nil, err
+		}
+
+		trustedHeader, ok = header.(*tmclient.Header)
+		if !ok {
+			return nil, fmt.Errorf("failed to check for misbehaviour, expected %T, got %T", (*tmclient.Header)(nil), header)
+		}
+	} else {
+		trustedHeader, err = cachedHeader.(TendermintIBCHeader).TMHeader()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if isMatchingTendermintConsensusState(proposedHeader.ConsensusState(), trustedHeader.ConsensusState()) {
+		return nil, nil
+	}
+
+	trustedHeader.TrustedValidators = proposedHeader.TrustedValidators
```

Comment:
did you figure out the reason for these two overrides? A comment would be helpful to indicate why

Reply:
stuck a comment in there now!